cranelift/codegen/src/ir/pcc.rs

```diff
@@ -0,0 +1,347 @@
+//! Proof-carrying code. We attach "facts" to values and then check
+//! that they remain true after compilation.
+//!
+//! A few key design principle of this approach are:
+//!
+//! - The producer of the IR provides the axioms. All "ground truth",
+//!   such as what memory is accessible -- is meant to come by way of
+//!   facts on the function arguments. In some sense, all we are doing
+//!   here is validating the "internal consistency" of the facts that
+//!   are provided on values, and the actions performed on those
+//!   values.
+//!
+//! - We do not derive and forward-propagate facts eagerly. Rather,
+//!   the producer needs to provide breadcrumbs -- a "proof witness"
+//!   of sorts -- to allow the checking to complete. That means that
+//!   as an address is computed, or pointer chains are dereferenced,
+//!   each intermediate value will likely have some fact attached.
+//!
+//!   This does create more verbose IR, but a significant positive
+//!   benefit is that it avoids unnecessary work: we do not build up a
+//!   knowledge base that effectively encodes the integer ranges of
+//!   many or most values in the program. Rather, we only check
+//!   specifically the memory-access sequences. In practice, each such
+//!   sequence is likely to be a carefully-controlled sequence of IR
+//!   operations from, e.g., a sandboxing compiler (such as
+//!   `cranelift-wasm`) so adding annotations here so communicate
+//!   intent (ranges, bounds-checks, and the like) is no problem.
+//!
+//! Facts are attached to SSA values in CLIF, and are maintained
+//! through optimizations and through lowering. They are thus also
+//! present on VRegs in the VCode. In theory, facts could be checked
+//! at either level, though in practice it is most useful to check
+//! them at the VCode level if the goal is an end-to-end verification
+//! of certain properties (e.g., memory sandboxing).
+//!
+//! Checking facts entails visiting each instruction that defines a
+//! value with a fact, and checking the result's fact against the
+//! facts on arguments and the operand. For VCode, this is
+//! fundamentally a question of the target ISA's semantics, so we call
+//! into the `LowerBackend` for this. Note that during checking there
+//! is also limited forward propagation / inference, but only within
+//! an instruction: for example, an addressing mode commonly can
+//! include an addition, multiplication/shift, or extend operation,
+//! and there is no way to attach facts to the intermediate values
+//! "inside" the instruction, so instead the backend can use
+//! `Fact::add()` and friends to forward-propagate facts.
+//!
+//! TODO:
+//! - Propagate facts through optimization (egraph layer).
+//! - Generate facts in cranelift-wasm frontend when lowering memory ops.
+//! - Implement richer "points-to" facts that describe the pointed-to
+//!   memory, so the loaded values can also have facts.
```

Comment:
I too had wondered if we could making global-value assertions axiomatic to avoid a richer assertion language so I'm glad you asked. I can definitely see downsides but it seems nice as a starting point.

Reply:
I've thought about this for a bit too; I think I'd still like to try actually describing memory layout (and I have some thoughts on how to do it while keeping `Fact` cheap with shared "memcaps" that are shared across the func). The basic reason is to avoid growing the trusted base. If we provide an axiom on what the loaded value from `vmctx` is, then we're trusting that load to match what we've stated; but the load-generation machinery is exactly what we're trying to verify with all of this in the first place. Now, arbitrary program logic may create more complex conditions / isel cases than a controlled `vmctx` case, but maybe not; in either case, I'd rather have the machine load instructions actually verified.